File: src/main/java/com/example/lee/thirdRound/TrappingRainWater.java
package com.example.lee.thirdRound;

import java.util.Stack;

/**
 * Created by benbendaisy on 7/7/17.
 */
public class TrappingRainWater {
    /**
     * O(n^2) solution, brute force and passed leet code
     * @param height
     * @return
     */
    public int trap(int[] height) {
        if (height == null || height.length < 2) {
            return 0;
        }
        int ans = 0;
        for (int i = 0; i < height.length; i++) {
            //left max
            int leftMax = 0;
            for (int j = 0; j <= i; j++) {
                leftMax = Math.max(leftMax, height[j]);
            }
            // right max
            int rightMax = 0;
            for (int j = height.length - 1; j >= i; j--) {
                rightMax = Math.max(rightMax, height[j]);
            }
            ans += Math.min(leftMax, rightMax) - height[i];
        }
        return ans;
    }

    /**
     * O(n) solution passed leet code
     * @param height
     * @return
     */
    public int trapI(int[] height) {
        if (height == null || height.length < 2) {
            return 0;
        }
        int[] leftMax = new int[height.length];
        int[] rightMax = new int[height.length];

        // find left maxes
        leftMax[0] = height[0];
        for (int i = 1; i < height.length; i++) {
            leftMax[i] = Math.max(leftMax[i - 1], height[i]);
        }

        // find right maxes
        rightMax[height.length - 1] = height[height.length - 1];
        for (int i = height.length - 2; i >= 0; i--) {
            rightMax[i] = Math.max(rightMax[i + 1], height[i]);
        }

        int res = 0;
        for (int i = 0; i < height.length; i++) {
            res += Math.min(leftMax[i], rightMax[i]) - height[i];
        }
        return res;
    }

    /**
     * O(n) solution with stack to save index of height
     * refer to: https://leetcode.com/articles/trapping-rain-water/
     * @param height
     * @return
     */
    public int trapII(int[] height) {
        if (height == null || height.length < 2) {
            return 0;
        }
        Stack<Integer> stack = new Stack<>();
        int res = 0;
        for (int i = 0; i < height.length; i++) {
            while (!stack.isEmpty() && height[stack.peek()] < height[i]) {
                int top = stack.peek();
                stack.pop();
                if (stack.isEmpty()) {
                    continue;
                }
                int dist = i - stack.peek() - 1;
                int h = Math.min(height[stack.peek()], height[i]) - height[top];
                res += dist * h;
            }
            stack.push(i);
        }
        return res;
    }

    public int trapIII(int[] height) {
        if (height == null || height.length < 2) {
            return 0;
        }
        int leftMax = 0, rightMax = 0;
        int l = 0, r = height.length - 1;
        int res = 0;
        while (l < r) {
            if (height[l] < height[r]) {
                if (leftMax > height[l]) {
                    res += leftMax - height[l];
                } else {
                    leftMax = height[l];
                }
                l++;
            } else {
                if (rightMax > height[r]) {
                    res += rightMax - height[r];
                } else {
                    rightMax = height[r];
                }
                r--;
            }
        }
        return res;
    }

}

File: python_module/examples/trapping_rain_water.py
from collections import deque
from typing import List


class Solution:
    def trap1(self, height: List[int]) -> int:
        if not height:
            return 0

        l, r = [0] * len(height), [0] * len(height)
        l[0], r[len(height) - 1] = height[0], height[len(height) - 1]

        for i in range(1, len(height)):
            l[i] = max(l[i - 1], height[i])
            r[-i] = max(r[-i + 1], height[-i])

        res = 0
        for i in range(1, len(height) - 1):
            res += min(l[i], r[i]) - height[i]

        return res

    def trap(self, height: List[int]) -> int:
        if not height:
            return 0

        q = deque()
        res = 0
        for cur in range(len(height)):
            while q and height[q[-1]] < height[cur]:
                h = height[q.pop()]
                if not q:
                    break
                res += (min(height[q[-1]], height[cur]) - h) * (cur - q[-1] - 1)
            q.append(cur)
        return res

if __name__ == "__main__":
    height = [4,2,0,3,2,5]
    solution = Solution()
    ret = solution.trap(height)
    print(ret)